### PR TITLE
Javadoc

### PR DESCRIPTION
**Pull request by rsavin**

### CODE REVIEW

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMConstants.java [hunk 1]
```diff
@@ -32,23 +32,23 @@
      */
     public static final String PROVIDER_NAME = "Google Cloud Messaging";
 
-    static final String C2DM_ACTION_RECEIVE = "com.google.android.c2dm.intent.RECEIVE";
+    public static final String GCM_NOT_CURRENT_PROVIDER_ERROR = "GCM is no current provider";
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMConstants.java]: no->not..

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMConstants.java [hunk 2]
```diff
@@ -32,23 +32,23 @@
...
+    public static final String REGISTRATION_ID_NOT_OBTAINED_ERROR = "Registration token isn't obtained";
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMConstants.java]: wasn't?

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMInstanceIDListenerService.java [hunk 3]
```diff
@@ -21,7 +21,8 @@
 import org.onepf.opfutils.OPFLog;
 
 /**
- * TODO javadoc
+ * The inheritor of the {@link InstanceIDListenerService}.
```
Comment (RomanZhilich) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMInstanceIDListenerService.java]: Seems kinds useles

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 4]
```diff
@@ -36,7 +36,8 @@
 import static org.onepf.opfpush.gcm.GCMConstants.REGISTRATION_ID_NOT_OBTAINED_ERROR;
 
 /**
- * TODO javadoc
+ * The helper class that perform {@link GcmPubSub#subscribe(String, String, Bundle)} and {@link GcmPubSub#unsubscribe(String, String)}
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: a helper class that performs

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 5]
```diff
@@ -71,6 +72,20 @@ public static GCMPubSubHelper getInstance(@NonNull final Context context) {
         return instance;
     }
 
+    /**
+     * Invokes {@link GcmPubSub#subscribe(String, String, Bundle)} asynchronously.
+     * Saved in {@link OPFPushHelper} registration id is used for the subscribing.
+     * It checks current provider name. If {@link OPFPushHelper#getProviderName()} isn't equal to {@link GCMConstants#PROVIDER_NAME}
+     * the {@link org.onepf.opfpush.gcm.GCMPubSubHelper.Callback#onError(String)} method will be called with
+     * {@link GCMConstants#GCM_NOT_CURRENT_PROVIDER_ERROR} as argument.
+     * Also it checks if the registration id is already obtained. If {@link OPFPushHelper#getRegistrationId()} returns null
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: can't parse

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 6]
```diff
@@ -96,6 +111,19 @@ public void run() {
         }
     }
 
+    /**
+     * Invokes {@link GcmPubSub#unsubscribe(String, String)} asynchronously.
+     * Saved in {@link OPFPushHelper} registration id is used for the unsubscribing.
+     * It checks current provider name. If {@link OPFPushHelper#getProviderName()} isn't equal to {@link GCMConstants#PROVIDER_NAME}
+     * the {@link org.onepf.opfpush.gcm.GCMPubSubHelper.Callback#onError(String)} method will be called with
+     * {@link GCMConstants#GCM_NOT_CURRENT_PROVIDER_ERROR} as argument.
+     * Also it checks if the registration id is already obtained. If {@link OPFPushHelper#getRegistrationId()} returns null
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: can't parse the same structure again(

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 7]
```diff
@@ -96,6 +111,19 @@ public void run() {
...
+     * the {@link org.onepf.opfpush.gcm.GCMPubSubHelper.Callback#onError(String)} method will be called with
+     * {@link GCMConstants#REGISTRATION_ID_NOT_OBTAINED_ERROR} as argument.
+     *
+     * @param topic    From which to stop receiving messages.
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: is it correct to say like "from which to stop"?

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 8]
```diff
@@ -96,6 +111,19 @@ public void run() {
         }
     }
 
+    /**
+     * Invokes {@link GcmPubSub#unsubscribe(String, String)} asynchronously.
+     * Saved in {@link OPFPushHelper} registration id is used for the unsubscribing.
```
Comment (RomanZhilich) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: Registration id saved in OPFPushHelper will be used for unsubscribing.

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 9]
```diff
@@ -36,7 +36,8 @@
 import static org.onepf.opfpush.gcm.GCMConstants.REGISTRATION_ID_NOT_OBTAINED_ERROR;
 
 /**
- * TODO javadoc
+ * The helper class that perform {@link GcmPubSub#subscribe(String, String, Bundle)} and {@link GcmPubSub#unsubscribe(String, String)}
+ * in the worker thread using saved registration id.
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: the saved?

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 10]
```diff
@@ -71,6 +72,20 @@ public static GCMPubSubHelper getInstance(@NonNull final Context context) {
         return instance;
     }
 
+    /**
+     * Invokes {@link GcmPubSub#subscribe(String, String, Bundle)} asynchronously.
+     * Saved in {@link OPFPushHelper} registration id is used for the subscribing.
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: do we need the here:
the subscribing?

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 11]
```diff
@@ -137,36 +175,80 @@ public String toString() {
         @Nullable
         private String clickAction;
 
+        /**
+         * Indicates notification title.
```
Comment (RomanZhilich) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: Indicates?

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 12]
```diff
@@ -71,6 +72,20 @@ public static GCMPubSubHelper getInstance(@NonNull final Context context) {
...
+     * It checks current provider name. If {@link OPFPushHelper#getProviderName()} isn't equal to {@link GCMConstants#PROVIDER_NAME}
+     * the {@link org.onepf.opfpush.gcm.GCMPubSubHelper.Callback#onError(String)} method will be called with
+     * {@link GCMConstants#GCM_NOT_CURRENT_PROVIDER_ERROR} as argument.
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: as THE(?) argument

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 13]
```diff
@@ -71,6 +72,20 @@ public static GCMPubSubHelper getInstance(@NonNull final Context context) {
...
+     * Also it checks if the registration id is already obtained. If {@link OPFPushHelper#getRegistrationId()} returns null
+     * the {@link org.onepf.opfpush.gcm.GCMPubSubHelper.Callback#onError(String)} method will be called with
+     * {@link GCMConstants#REGISTRATION_ID_NOT_OBTAINED_ERROR} as argument.
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: as THE(?) argument

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 14]
```diff
@@ -71,6 +72,20 @@ public static GCMPubSubHelper getInstance(@NonNull final Context context) {
...
+     *
+     * @param topic    Developer defined topic name. Must match the following regular expression: "/topics/[a-zA-Z0-9-_.~%]{1,900}".
+     * @param extras   (optional) additional information.
+     * @param callback Callback instance. Can be null.
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: THE(?) callback instance

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 15]
```diff
@@ -131,7 +159,14 @@ private boolean checkPubSubAvailability(@NonNull final OPFPushHelper helper, @No
         return true;
     }
 
-    public class CallbackMainThreadWrapper implements Callback {
+    public interface Callback {
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: do we need such abstract thing here? can we move it..say, to the utils?

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java [hunk 16]
```diff
@@ -131,7 +159,14 @@ private boolean checkPubSubAvailability(@NonNull final OPFPushHelper helper, @No
...
+
+        void onSuccess();
+
+        void onError(@Nullable final String error);
+    }
+
+    private class CallbackMainThreadWrapper implements Callback {
```
Comment (akarimova) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: do we need such abstract thing here? can we move it..say, to the utils?

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 17]
```diff
@@ -66,36 +67,73 @@ private NotificationPayload(@NonNull final String title,
...
+     *
+     * @return The field that indicates notification title.
+     */
     @NonNull
     public String getTitle() {
         return title;
     }
 
+    /**
+     * Returns the field that indicates notification icon.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the notification icon

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 18]
```diff
@@ -66,36 +67,73 @@ private NotificationPayload(@NonNull final String title,
...
+     *
+     * @return The field that indicates notification icon.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the notification icon

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 19]
```diff
@@ -66,36 +67,73 @@ private NotificationPayload(@NonNull final String title,
...
+     */
     @NonNull
     public String getIcon() {
         return icon;
     }
 
+    /**
+     * Returns the field that indicates notification body text.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the notification body

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 20]
```diff
@@ -66,36 +67,73 @@ private NotificationPayload(@NonNull final String title,
...
+     *
+     * @return The field that indicates notification body text.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the notification body

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 21]
```diff
@@ -66,36 +67,73 @@ private NotificationPayload(@NonNull final String title,
...
+     */
     @Nullable
     public String getBody() {
         return body;
     }
 
+    /**
+     * Returns the field that indicates sound to be played.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a(the) sound

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 22]
```diff
@@ -66,36 +67,73 @@ private NotificationPayload(@NonNull final String title,
...
+     *
+     * @return The field that indicates sound to be played. Supports only {@code default} currently.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a(the) sound
Supports only {@code default} currently. - can't parse

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 23]
```diff
@@ -66,36 +67,73 @@ private NotificationPayload(@NonNull final String title,
...
+     */
     @Nullable
     public String getSound() {
         return sound;
     }
 
+    /**
+     * Indicates whether each notification message results in a new entry on the notification center on Android.
+     * If not set, each request creates a new notification. If set, and a notification with the same tag is already being shown,
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: not sure about comma

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 24]
```diff
@@ -66,36 +67,73 @@ private NotificationPayload(@NonNull final String title,
...
+     *
+     * @return The field that indicates color of the icon, expressed in #rrggbb format.
+     */
     @Nullable
     public String getColor() {
         return color;
     }
 
+    /**
+     * The action associated with a user click on the notification.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: the user

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 25]
```diff
@@ -66,36 +67,73 @@ private NotificationPayload(@NonNull final String title,
...
+     *
+     * @return The action associated with a user click on the notification.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: the

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java (hunk 11 above)
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 26]
```diff
@@ -137,36 +175,80 @@ public String toString() {
         @Nullable
         private String clickAction;
 
+        /**
+         * Indicates notification title.
+         *
+         * @param title Notification title. Required parameter.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 27]
```diff
@@ -137,36 +175,80 @@ public String toString() {
...
+         * @return Builder instance.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: the

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 28]
```diff
@@ -137,36 +175,80 @@ public String toString() {
...
+         */
         public Builder setTitle(@NonNull final String title) {
             this.title = title;
             return this;
         }
 
+        /**
+         * Indicates notification icon. Sets value to myicon for drawable resource myicon.png. Required parameter.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 29]
```diff
@@ -137,36 +175,80 @@ public String toString() {
...
+         * @param icon Notification icon. Required parameter.
+         * @return Builder instance.
+         */
         public Builder setIcon(@NonNull final String icon) {
             this.icon = icon;
             return this;
         }
 
+        /**
+         * Indicates notification body text.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 30]
```diff
@@ -137,36 +175,80 @@ public String toString() {
...
+         *
+         * @param body Notification body text. Optional parameter.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 31]
```diff
@@ -137,36 +175,80 @@ public String toString() {
...
+         * @return Builder instance.
+         */
         public Builder setBody(@Nullable final String body) {
             this.body = body;
             return this;
         }
 
+        /**
+         * Indicates sound to be played. Supports only default currently.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: a/the

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 32]
```diff
@@ -137,36 +175,80 @@ public String toString() {
...
+         *
+         * @param sound Notification sound. Optional parameter.
+         * @return Builder instance.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: the

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 33]
```diff
@@ -137,36 +175,80 @@ public String toString() {
...
+         */
         public Builder setSound(@Nullable final String sound) {
             this.sound = sound;
             return this;
         }
 
+        /**
+         * Indicates whether each notification message results in a new entry on the notification center on Android.
+         * If not set, each request creates a new notification. If set, and a notification with the same tag is already
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: not sure about comma

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java [hunk 34]
```diff
@@ -137,36 +175,80 @@ public String toString() {
...
+         * @param tag Notification tag. Optional parameter.
+         * @return Builder instance.
+         */
         public Builder setTag(@Nullable final String tag) {
             this.tag = tag;
             return this;
         }
 
+        /**
+         * Indicates color of the icon, expressed in #rrggbb format.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: not sure about comma

File: opfpush/src/main/java/org/onepf/opfpush/notification/OPFNotificationMaker.java [hunk 35]
```diff
@@ -36,20 +38,43 @@
...
+     *                             to a {@link org.onepf.opfpush.model.NotificationPayload} instance.
+     *                             It can be useful for implementing custom converting from {@link Bundle}
+     *                             to {@link org.onepf.opfpush.model.NotificationPayload}.
+     */
     public OPFNotificationMaker(@NonNull final NotificationPreparer notificationPreparer) {
         this.notificationPreparer = notificationPreparer;
     }
 
+    /**
+     * Returns {@code true} if the {@code bundle} contains an extra by the "opf_notification" key and it is "true" String.
```
Comment (akarimova) [inline opfpush/src/main/java/org/onepf/opfpush/notification/OPFNotificationMaker.java]: "extra by" is it ok?

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java (hunk 11 above)
Comment (rsavin) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: Copypaste from gcm docs https://developers.google.com/cloud-messaging/server-ref#notification-payload-support

File: opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java (hunk 15 above)
Comment (rsavin) [inline opfpush-providers/gcm/src/main/java/org/onepf/opfpush/gcm/GCMPubSubHelper.java]: This interface should be used only for GCMPubSubHelper methods therefore it is inner interface. There are a lot of inner Callback interfaces in the android SDK.

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java (hunk 17 above)
Comment (rsavin) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: Copypaste of gcm docs

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java (hunk 18 above)
Comment (rsavin) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: Copypaste of gcm docs

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java (hunk 19 above)
Comment (rsavin) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: Copypaste of gcm docs

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java (hunk 20 above)
Comment (rsavin) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: Copypaste of gcm docs

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java (hunk 21 above)
Comment (rsavin) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: Copypaste of gcm docs

File: opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java (hunk 22 above)
Comment (rsavin) [inline opfpush/src/main/java/org/onepf/opfpush/model/NotificationPayload.java]: >> can't parse
It's copypaste too. It's mean that you can put only "default" string as parameter. Other values will be ignored.

**Pull request merged**